Refactor editor http server implementation

vlaaad opened the pull request:
Now editor scripts may add routes to the built-in HTTP server, e.g.:
```lua
function M.get_server_routes()
  return {
    http.server.route("/my-extension/files/{*file}", function(request)
      local attrs = editor.external_file_attributes(request.file)
      if attrs.is_file then
        return http.server.external_file_response(request.file)
      else
        return http.server.response(404, "Not found")
      end
    end)
  }
end
```

Fixes #10272

vlaaad committed:
Refactor editor http server implementation

This will allow us to expose it to editor scripts without too much pain for the users.

Related to #10272

vlaaad commented on editor/project.clj:
```diff
@@ -63,6 +63,8 @@
                      [com.defold.lib/bob                          "1.0"]
                      [com.defold.lib/openmali                     "1.0"]
 
+                     [metosin/reitit-core "0.8.0-alpha1"]
```
Router library. This dependency increases the editor size by 33kb.

vlaaad commented on editor/test/integration/web_server_test.clj:
```diff
@@ -0,0 +1,66 @@
+                (is (= 202 status))
+                (is (= "202 Accepted\n" body)))
+              (when (is (realized? ui/open-url))
+                (is (string/includes? @ui/open-url "github.com/defold/defold/issues"))))))))))
```
I added a test for the other endpoints of the HTTP server besides the hot-reload ones tested in `hot-reload-test`

matgis commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+                          (.sendResponseHeaders
+                            exchange
+                            status
+                            ;; Response length argument:
+                            ;; -1 means empty response body (0 bytes)
+                            ;; 0 is unknown size (chunked)
+                            ;; positive is exact size
+                            (if (nil? connection)
+                              -1
+                              (if-let [explicit-length (get headers "content-length")]
+                                (let [n (Long/parseUnsignedLong explicit-length)]
+                                  (if (zero? n) -1 n))
+                                0)))
```
Really annoying for the `responseLength` argument to behave like this. 😬 I would have expected the zero and -1 cases to be swapped, if anything.

matgis commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+                                  ;; the connection before we finished sending
+                                  ;; the response: ignore such exceptions since
+                                  ;; it's not an issue.
+                                  (catch IOException _)))
```
According to the docs for `getResponseBody`:

> If too many bytes are written, then the write method of `OutputStream` will throw an `IOException`. If too few bytes are written then the stream `OutputStream.close()` will throw an `IOException`.

I think the case you're catching here is when _too many_ bytes are written to the output stream returned by `.getResponseBody`? `io/copy` doesn't close the stream, so presumably that won't happen until `.close` is called on the `exchange` in the `finally` clause below?

vlaaad replied:
Hmm, that's a good point. The actual exceptions I was getting were `Broken pipe` and `Connection reset by peer` when I was holding `F5` in the browser, which caused it to constantly reload the engine profiler page. I think I'll log the exceptions instead of ignoring them completely (in another PR)

matgis commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+                                method-set (cond-> method-set (contains? method-set "GET") (conj "HEAD"))]
+                            (string/join ", " (sort method-set))))
+        invoke-handler (fn invoke-handler [handler request match]
+                         (handler (assoc request :path-params (:path-params match))))
```
These functions don't appear to use anything from the `routes`, so they could be global?

vlaaad replied:
Yes, I'll make them global (in another PR)

matgis commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+    (fn route-request [request]
+      (let [path (:path request)]
+        (if-let [match (or (reitit/match-by-path router path)
+                           (if (string/ends-with? path "/")
+                             (reitit/match-by-path router (subs path 0 (dec (count path))))
+                             (reitit/match-by-path router (str path "/"))))]
```
In their own examples, they do a redirect when this happens, possibly to ensure the client ends up at the corrected url. Not sure if we want to do the same? 🤔

See https://cljdoc.org/d/metosin/reitit/0.5.18/doc/ring/slash-handler

matgis commented on editor/test/integration/editor_extensions_test.clj:
```diff
@@ -855,16 +854,14 @@ POST http://localhost:23456/echo hello world! as string => 200
+                     {"/redirect/foo" {"GET" (constantly (http-server/redirect "/foo"))}
+                      "/foo" {"GET" (constantly (http-server/response 200 "successfully redirected"))}
+                      "/" {"GET" (constantly (http-server/response 200 ""))}
+                      "/json" {"GET" (constantly (http-server/response 200 "{\"a\": 1, \"b\": [true]}"))}
```
Should we use `http-server/json-response` here?

matgis commented on editor/test/integration/test_util.clj:
```diff
@@ -793,33 +793,34 @@
-
-(defn kill-lib-server [server]
-  (http-server/stop! server))
+(defmacro with-server [handler & body]
```
Rename to `with-http-server`? This is a bit generic for `test-util`. There can be other servers eventually, I guess?

vlaaad replied:
Heh, I thought about it but didn't come up with other servers we might have. Maybe we could do it when we actually have another non-http server? 😄

matgis replied:
We could, but I prefer we do it now for clarity. We're a desktop application, so "server" could mean other things (a build server or native extension server maybe?).

I'd especially like to address this now since it has an entry in `codeStyles/Project.xml`, so a refactoring rename will mess up the indentation if we're not careful.

matgis commented on editor/test/integration/test_util.clj:
```diff
@@ -793,33 +793,34 @@
+
+(def lib-server-handler
+  (http-server/router-handler
+    {"/lib/{*lib}" {"GET" (fn [request]
```
Break up into multiple lines to control indentation?

matgis commented on editor/test/integration/web_server_test.clj:
```diff
@@ -0,0 +1,66 @@
+  (:import [javafx.scene Scene]
+           [javafx.scene.layout Region]
+           [javafx.stage Stage]))
+
```
Add missing copyright header and `(set! *warn-on-reflection* true)`.

matgis reviewed (changes requested):
Stellar effort! It's a lot nicer after this. ❤️

I've not done any manual testing of this, because it seems like a lot of surface area to cover. I'm gonna rely on the automated tests.

I'm also not very knowledgeable when it comes to this stuff, but I think I may have found some concerns. I'll let you have a look and respond! 😅

vlaaad commented on editor/src/clj/editor/hot_reload.clj:
```diff
@@ -91,16 +44,24 @@
+                                          (fs/path-exists? full-path)
+                                          (not (fs/path-is-directory? full-path)))
+                                   (let [etag (workspace/etag workspace (str "/" path))
+                                         remote-etag (get headers "if-none-match")]
```
Yes, exactly. I ensure lower-case headers (both in the server and client!)

vlaaad commented on editor/src/clj/editor/html_view.clj:
```diff
@@ -25,66 +25,35 @@
+                                (http-server/response 200 resource)))
+                            (do (log/warn :message (format "Cannot find resource for %s" path))
+                                http-server/not-found)))}})
+    :host "0.0.0.0"))
```
Oops, I didn't notice it was `127.0.0.1` and not `0.0.0.0`. I guess there is no difference between these in this case, but it is important to provide a host here; otherwise, clicking on links that point to other in-project HTML resources won't work: it's used to make `(string/starts-with? new-location url)` work in `handle-location-change!`.

vlaaad commented on editor/src/clj/editor/resource.clj:
```diff
@@ -326,7 +333,26 @@
+  (content-type [this] (http-server/ext->content-type (type-ext this)))
+  http-server/DataConnection
+  (data-connection [_]
+    (let [zip-file (ZipFile. ^File (io/as-file zip-uri))
```
Good point, I'll change it.

vlaaad commented on editor/src/clj/editor/resource.clj:
```diff
@@ -326,7 +333,26 @@
+        (make-output-stream [_ _] (throw (Exception. "Zip resources are read-only")))
+        (make-writer [_ _] (throw (Exception. "Zip resources are read-only")))
+        AutoCloseable
+        (close [_] (.close zip-file))))))
```
Yes, because the entry input stream is the only input stream we return when opening the file here. FYI, this code mostly repeats the `make-zip-resource-input-stream` fn that we use for reading zip resources but does so using a reify instead of a proxy.

vlaaad commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+
+(extend-protocol DataLength
+  byte/1 (data-length [bytes] (count bytes))
+  Object (data-length [_])
```
No, it would not. These protocols define an open extension point for inferring content length and type for inputs, but there are still other valid bodies (i.e., anything implementing `io/IOFactory`) that can't report their content length. For example, you can make a response with an `InputStream` instance as a body — the HTTP server will successfully write it to the output using chunked transfer encoding. We don't want any errors in this case.

vlaaad commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+           (make-input-stream [_ opts] (io/make-input-stream path opts))
+           (make-output-stream [_ opts] (io/make-output-stream path opts))
+           (make-reader [_ opts] (io/make-reader path opts))
+           (make-writer [_ opts] (io/make-writer path opts))))
```
Yes, I agree.

vlaaad commented on editor/src/clj/util/http_server.clj:
```diff
@@ -13,98 +13,415 @@
+            (close [_]
+              (.close (.getInputStream connection))))))
+  URI (data-connection [uri] (data-connection (.toURL uri)))
+  Object (data-connection [x] x)
```
No, this enables automatic support for all the other `io/IOFactory` implementations like `InputStream` instances.

vlaaad committed:
Review fixes

vlaaad committed:
Add editor HTTP server extensibility (#10314)

* Add editor HTTP server extensibility

Now editor scripts may add routes to the built-in HTTP server, e.g.:
```lua
function M.get_server_routes()
  return {
    http.server.route("/my-extension/files/{*file}", function(request)
      local attrs = editor.external_file_attributes(request.file)
      if attrs.is_file then
        return http.server.external_file_response(request.file)
      else
        return http.server.response(404, "Not found")
      end
    end)
  }
end
```

Technical notes:

Doc PR: https://github.com/defold/doc/pull/529

Fixes #10272

* Add regex coercer test
